File: android/src/main/kotlin/com/example/cook_lib/CookLibPlugin.kt
package com.example.cook_lib

import android.content.Context
import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import java.util.concurrent.Executors

class CookLibPlugin: FlutterPlugin, MethodCallHandler, EventChannel.StreamHandler {
    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private lateinit var context: Context
    private var eventSink: EventChannel.EventSink? = null
    private val executor = Executors.newSingleThreadExecutor()

    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext

        // 音频解码通道 - 用于一次性解码请求
        methodChannel = MethodChannel(flutterPluginBinding.binaryMessenger, "cook_lib/audio/decoder")
        methodChannel.setMethodCallHandler(this)

        // 视频帧流通道 - 用于流式帧数据传输
        eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "cook_lib/video/frame_stream")
        eventChannel.setStreamHandler(this)
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        when (call.method) {
            "decodeToWav" -> {
                val inputPath = call.argument<String>("inputPath")
                if (inputPath == null) {
                    result.error("INVALID_ARGUMENT", "inputPath is required", null)
                    return
                }
                executor.execute {
                    try {
                        val wavPath = AudioDecoder.decodeToWav(context, inputPath)
                        android.os.Handler(android.os.Looper.getMainLooper()).post {
                            result.success(wavPath)
                        }
                    } catch (e: Exception) {
                        android.os.Handler(android.os.Looper.getMainLooper()).post {
                            result.error("DECODE_ERROR", e.message, null)
                        }
                    }
                }
            }
            else -> {
                result.notImplemented()
            }
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events

        // 从参数中获取 videoPath，立即开始提取
        val args = arguments as? Map<*, *>
        val videoPath = args?.get("videoPath") as? String
        if (videoPath != null) {
            executor.execute {
                VideoFrameExtractor.extractFrames(videoPath, eventSink)
            }
        }
    }

    override fun onCancel(arguments: Any?) {
        // 取消订阅时自动停止提取
        VideoFrameExtractor.stopExtraction()
        eventSink = null
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        executor.shutdown()
    }
}

File: ios/Classes/CookLibPlugin.swift
import Flutter
import UIKit

public class CookLibPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
    private var eventSink: FlutterEventSink?
    private let queue = DispatchQueue(label: "com.example.cook_lib.decoder", qos: .userInitiated)

    public static func register(with registrar: FlutterPluginRegistrar) {
        // 音频解码通道 - 用于一次性解码请求
        let audioChannel = FlutterMethodChannel(name: "cook_lib/audio/decoder", binaryMessenger: registrar.messenger())
        // 视频帧流通道 - 用于流式帧数据传输
        let videoFrameChannel = FlutterEventChannel(name: "cook_lib/video/frame_stream", binaryMessenger: registrar.messenger())

        let instance = CookLibPlugin()
        registrar.addMethodCallDelegate(instance, channel: audioChannel)
        videoFrameChannel.setStreamHandler(instance)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "decodeToWav":
            guard let args = call.arguments as? [String: Any],
                  let inputPath = args["inputPath"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "inputPath is required", details: nil))
                return
            }

            queue.async {
                do {
                    let wavPath = try AudioDecoder.decodeToWav(inputPath: inputPath)
                    DispatchQueue.main.async {
                        result(wavPath)
                    }
                } catch {
                    DispatchQueue.main.async {
                        result(FlutterError(code: "DECODE_ERROR", message: error.localizedDescription, details: nil))
                    }
                }
            }

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - FlutterStreamHandler

    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events

        // 从参数中获取 videoPath，立即开始提取
        if let args = arguments as? [String: Any],
           let videoPath = args["videoPath"] as? String {
            queue.async { [weak self] in
                VideoFrameExtractor.extractFrames(videoPath: videoPath, eventSink: self?.eventSink)
            }
        }

        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        // 停止当前实例的提取
        VideoFrameExtractor.stopExtraction()
        self.eventSink = nil
        return nil
    }
}
